calyx-opt/src/passes/compile_static.rs

```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
-            .map(|assign| {
-                make_assign_dyn(assign, &g, &fsm, fsm_size, &mut builder)
-            })
+            .map(|assign| make_assign_dyn(assign, &fsm, fsm_size, builder))
             .collect_vec();
         // assignments to increment the fsm
-        let not_last_state_guard: ir::Guard<ir::Nothing> =
-            guard!(fsm["out"]).neq(guard!(last_state["out"]));
-        let last_state_guard: ir::Guard<ir::Nothing> =
-            guard!(fsm["out"]).eq(guard!(last_state["out"]));
+        let not_penultimate_state_guard: ir::Guard<ir::Nothing> =
+            guard!(fsm["out"]).neq(guard!(penultimate_state["out"]));
+        let penultimate_state_guard: ir::Guard<ir::Nothing> =
+            guard!(fsm["out"]).eq(guard!(penultimate_state["out"]));
         let fsm_incr_assigns = build_assignments!(
           builder;
+          // increments the fsm
           adder["left"] = ? fsm["out"];
           adder["right"] = ? const_one["out"];
-          fsm["write_en"] = not_last_state_guard ? signal_on["out"];
-          fsm["in"] = not_last_state_guard ? adder["out"];
-          // need done condition because we are creating a dynamic group
-          g["done"] = last_state_guard ? signal_on["out"];
+          fsm["write_en"] = ? signal_on["out"];
+          fsm["in"] = not_penultimate_state_guard ? adder["out"];
+           // resets the fsm early
+          fsm["in"] = penultimate_state_guard ? first_state["out"];
+          // will never reach this guard since we are resetting when we get to
+          // the penultimate state
+          g["done"] = ? ud["out"];
         );
         assigns.extend(fsm_incr_assigns.to_vec());
+        // maps the "early reset" group name to the "fsm name" that it borrows.
+        // this is helpful when we build the "wrapper group"
+        self.fsm_map.insert(g.borrow().name(), fsm.borrow().name());
         // adding the assignments to the new dynamic group and creating a
         // new (dynamic) enable
         g.borrow_mut().assignments = assigns;
-        g.borrow_mut().attributes = sgroup.attributes.clone();
-        let mut e = ir::Control::enable(g);
-        let attrs = std::mem::take(&mut s.attributes);
-        *e.get_mut_attributes() = attrs;
-        // need to add a continuous assignment to reset the fsm
-        let fsm_reset_assigns = build_assignments!(builder;
-            fsm["in"] = last_state_guard ? first_state["out"];
-            fsm["write_en"] = last_state_guard ? signal_on["out"];
+        g.borrow_mut().attributes = attributes;
+        g
+    }
+
+    fn build_wrapper_group(
+        fsm_name: &ir::Id,
+        group_name: &ir::Id,
+        builder: &mut ir::Builder,
+    ) -> ir::RRC<ir::Group> {
+        // get the groups/fsm necessary to build the wrapper group
+        let early_reset_group = builder
+            .component
+            .get_groups()
+            .find(*group_name)
+            .unwrap_or_else(|| {
+                panic!(
```

Comment:
Per @EclecticGriffin's recommendation, avoid using `panic!` directly. If this case should never occur, use `unreachable!` instead